src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,363 @@
+        }
+
+        /// <inheritdoc />
+        public async Task InitializeAsync()
+        {
+            await this.provenBlockHeaderRepository.InitializeAsync().ConfigureAwait(false);
+
+            this.TipHashHeight = this.provenBlockHeaderRepository.TipHashHeight;
+
+            this.logger.LogDebug("Initialized ProvenBlockHader block tip at '{0}'.", this.TipHashHeight);
+
+            this.asyncLoop = this.asyncLoopFactory.Run("ProvenBlockHeaders job", token =>
+            {
+                // Save pending items.
+                this.SaveAsync().ConfigureAwait(false);
+
+                // Check and make sure the cache limit hasn't been breached.
+                this.ManangeCacheSize();
+
+                this.logger.LogTrace("(-)[IN_ASYNC_LOOP]");
+
+                return Task.CompletedTask;
+            },
+            this.nodeLifetime.ApplicationStopping,
+            repeatEvery: TimeSpan.FromMinutes(1),
+            startAfter: TimeSpan.FromMinutes(1));
+        }
+
+        /// <inheritdoc />
+        public async Task<ProvenBlockHeader> GetAsync(int blockHeight)
+        {
+            ProvenBlockHeader header = null;
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddQueryTime(o)))
+            {
+                header = this.CheckCache(blockHeight);
+
+                if (header == null)
+                {
+                    // Check the repository (DBreeze).
+                    header = await this.provenBlockHeaderRepository.GetAsync(blockHeight).ConfigureAwait(false);
+
+                    this.AddToCache(blockHeight, header);
+                }
+            }
+
+            return header;
+        }
+
+        /// <inheritdoc />
+        public async Task<List<ProvenBlockHeader>> GetAsync(int fromBlockHeight, int toBlockHeight)
+        {
+            var headersInCache = new List<ProvenBlockHeader>();
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddQueryTime(o)))
+            {
+                int heightKey = fromBlockHeight;
+
+                ProvenBlockHeader header = null;
+
+                var headersNotInCache = new List<int>();
+
+                do
+                {
+                    header = this.CheckCache(heightKey);
+
+                    if (header != null)
+                        headersInCache.Add(header);
+                    else
+                        headersNotInCache.Add(heightKey);
+
+                    heightKey++;
+
+                } while (heightKey <= toBlockHeight);
+
+                // Try and get items from the repository if not found in the store cache.
+                if (headersNotInCache.Count > 0)
+                {
+                    var repositoryHeaders = new List<ProvenBlockHeader>();
+
+                    foreach(int headerNotInCache in headersNotInCache)
+                    {
+                        var repositoryHeader = new ProvenBlockHeader();
```

Comment:
instead of this and next line you can do just

```
ProvenBlockHeader repositoryHeader = await this.provenBlockHeaderRepository.GetAsync(headerNotInCache).ConfigureAwait(false);
```